File: scripts/log.js
const puppeteer = require('puppeteer');

const args = process.argv.slice(2);

const FIREFOX_WS_ENDPOINT = 'ws://127.0.0.1:PORT/session'.replace('PORT', args[0]);

// Timestamp helper
const log = (label, info = '') => {
    const ts = new Date().toISOString();
    console.log(`[${ts}] [${label}] ${info}`);
};

const monitorPage = async (page) => {
    page.on('close', () => log('page.close', 'Page was closed.'));
    page.on('console', msg => log('page.console', `Message: ${msg.text()}`));
    page.on('dialog', dialog => {
        log('page.dialog', `Type: ${dialog.type()} | Message: ${dialog.message()}`);
        dialog.dismiss().catch(() => {}); // Avoid hanging if not handled
    });
    page.on('domcontentloaded', () => log('page.domcontentloaded', 'DOM fully loaded.'));
    page.on('error', err => log('page.error', `Page crashed or error occurred: ${err}`));
    page.on('frameattached', frame => log('page.frameattached', `URL: ${frame.url()}`));
    page.on('framedetached', frame => log('page.framedetached', `Frame detached.`));
    page.on('framenavigated', frame => log('page.framenavigated', `Navigated to: ${frame.url()}`));
    page.on('load', () => log('page.load', 'Page fully loaded.'));
    page.on('metrics', data => log('page.metrics', `Metrics reported: ${JSON.stringify(data)}`));
    page.on('pageerror', err => log('page.pageerror', `Unhandled exception: ${err}`));
    page.on('popup', popup => log('page.popup', 'A new popup/tab was opened.'));
    page.on('request', req => log('page.request', `URL: ${req.url()}`));
    page.on('requestfailed', req => log('page.requestfailed', `URL: ${req.url()} | Reason: ${req.failure().errorText}`));
    page.on('requestfinished', req => log('page.requestfinished', `URL: ${req.url()}`));
    page.on('response', res => log('page.response', `URL: ${res.url()} | Status: ${res.status()}`));
    page.on('workercreated', worker => log('page.workercreated', `Worker URL: ${worker.url()}`));
    page.on('workerdestroyed', worker => log('page.workerdestroyed', `Worker URL: ${worker.url()}`));
}

(async () => {
    console.log(`[+] Attempting to connect to: ${FIREFOX_WS_ENDPOINT}`);
    let browser = null;

    try {
        // Connect to an existing instance of Firefox via WebSocket
        browser = await puppeteer.connect({
            browserWSEndpoint: FIREFOX_WS_ENDPOINT,
            product: 'firefox',
            protocol: 'webDriverBiDi', // Required for Firefox >= 2023 versions
            headers: {
                Host: '127.0.0.1:9222', // Required to make the BiDi inside the container happy
            }
        });

        process.on('exit', async () => {
            console.log("[~] Disconnecting from Firefox...");
            await browser.disconnect();
            console.log('[+] Done!');
        });

        console.log('[+] Successfully connected to Firefox!');

        // Monitor existing pages
        pages = await browser.pages();
        for(const page of pages) {
            await monitorPage(page);
        }

        // Monitor new pages that are opened
        browser.on("targetcreated", async (target) => {
            if (target.type() === "page") {
                try {
                    const page = await target.page();
                    if (page) {
                        monitorPage(page)
                    }
                } catch (error) {
                    console.error(`Error processing new page: ${error.message}`);
                }
            }
        });

        // Monitor other browser events
        browser.on('targetdestroyed', target => {
            log('browser.targetdestroyed', `Target destroyed: ${target.url()}`);
        });

        browser.on("targetchanged", async (target) => {
            log('browser.targetchanged', `Target changed: ${target.url()}`);
        });
    } catch (error) {
        console.error('[X] Failed to connect or interact with the existing Firefox instance:', error);

        if (browser && browser.isConnected()) {
            try {
                await browser.disconnect();
            } catch (disconnectError) {
                console.error('[!] Error while trying to disconnect after failure:', disconnectError);
            }
        }
    }
})();

File: scripts/log.py
from patchright.async_api import async_playwright
import asyncio
from json import dumps
import sys

if len(sys.argv) < 2:
    print('[-] Usage: python3 log.py <DEBUGPORT>')
    sys.exit(1)
else:
    PORT = int(sys.argv[1])

CDP_ENDPOINT = f"http://127.0.0.1:{PORT}"

def print_event(event_name, payload):
    print(f"\n-- [{event_name}] --")
    try:
        if isinstance(payload, str):
            print(payload)
        elif hasattr(payload, 'url'):
            print("[+] URL:", payload.url)
        elif hasattr(payload, 'message'):
            print("[+] Message:", payload.message)
        elif hasattr(payload, 'type'):
            print("[+] Type:", payload.type)
        else:
            print(json.dumps(payload, indent=2, default=str))
    except Exception as e:
        print("[-] Error printing event:", e)


async def hook_page_events(page):
    # From ChatGPT, don't judge if it's wrong :-)
    events = [
        "close", "console", "crash", "dialog", "domcontentloaded", "download", "filechooser",
        "framenavigated", "load", "pageerror", "popup", "request", "requestfailed",
        "requestfinished", "response", "websocket", "websocketclosed", "websocketerror",
        "websocketframeerror", "websocketframereceived", "websocketframesent", "worker"
    ]

    for event in events:
        try:
            page.on(event, lambda payload, e=event: print_event(e, payload))
        except Exception as ex:
            print(f"[-] Failed to hook event {event}: {ex}")

    print(f"[+] Hooked events on page: {page.url}")

async def main():
    async with async_playwright() as p:
        print(f'[+] Attempting to connect to: {CDP_ENDPOINT}');
        try:
            browser = await p.chromium.connect_over_cdp(CDP_ENDPOINT)
            print('[+] Successfully connected to Chrome!');
        except Exception as e:
            print(f'[-] Error while connecting to the browser: {e}')


        context = browser.contexts[0] if browser.contexts else await browser.new_context()

        # Hook existing pages
        for page in context.pages:
            await hook_page_events(page)

        # Hook future pages
        context.on("page", hook_page_events)

        print("[+] Listening to all events on all pages. Press Ctrl+C to stop.")
        await asyncio.Event().wait()  # Keep running

asyncio.run(main())
